draco/asp/hard.lp

```diff
@@ -1,24 +1,97 @@
+violation(bin_q_o) :-
+    attribute(scale_type,M,C,T),
+    entity(encoding,M,E),
+    attribute((encoding,channel),E,C),
+    attribute((encoding,binning),E,_),
+    T != linear,
+    T != log,
+    T != ordinal.
+
+% @hard(log_non_positive) Cannot use log if the data is negative or zero.
+violation(log_non_positive) :-
+    attribute(extent,N,MIN,_),
+    attribute(channel_field,M,C,N),
+    attribute(scale_type,M,C,log),
+    MIN <= 0.
 
 % @hard(bin_and_aggregate) Cannot bin and aggregate.
 violation(bin_and_aggregate) :-
     attribute((encoding,binning),E,_),
     attribute((encoding,aggregate),E,_).
 
-% @hard(no_encodings) There has to be at least one encoding for every mark.
-violation(no_encodings) :-
-    entity(mark,_,M),
-    not entity(encoding,M,_).
+% @hard(aggregate_o_valid) Oridnal only supports min, max, and median.
+violation(aggregate_o_valid) :-
+    entity(encoding,M,E),
+    attribute((encoding,channel),E,C),
+    attribute((encoding,aggregate),E,A),
+    attribute(scale_type,M,C,ordinal),
+    A != min,
+    A != max,
+    A != median.
+
+% @hard(aggregate_t_valid) Temporal only supports min and max.
+violation(aggregate_t_valid) :-
+    attribute(field_type,N,datetime),
+    attribute((encoding,field),E,N),
+    attribute((encoding,aggregate),E,A),
+    A != min,
+    A != max.
+
+% @hard(aggregate_nominal) Cannot aggregate categorical.
+violation(aggregate_nominal) :-
+    entity(encoding,M,E),
+    attribute((encoding,channel),E,C),
+    attribute((encoding,aggregate),E,_),
+    attribute(scale_type,M,C,categorical).
+
+% @hard(aggregate_detail) Detail cannot be aggregated.
+violation(aggregate_detail) :-
+    attribute((encoding,channel),E,detail),
+    attribute((encoding,aggregate),E,_).
+
+% @hard(count_q_without_field) Count has to be quantitative and not use a field.
+violation(count_q_without_field) :-
+    attribute((encoding,field),E,_),
+    attribute((encoding,aggregate),E,count).
+violation(count_q_without_field) :-
+    entity(encoding,M,E),
+    attribute((encoding,aggregate),E,count),
+    attribute((encoding,channel),E,C),
+    attribute(scale_type,M,C,T),
+    T != linear,
+    T != log.
+
+% @hard(shape_discrete_non_ordered) Shape requires discrete and not ordered (nominal). Using ordinal would't make a difference in Vega-Lite.
+violation(shape_discrete_non_ordered) :-
+    attribute((scale,channel),E,shape),
+    not attribute((scale,type),E,categorical).
+
+% @hard(detail_non_ordered) Detail requires nominal.
+violation(detail_non_ordered) :-
+    attribute((scale,channel),E,detail),
+    not attribute((scale,type),E,categorical).
+
+% @hard(size_nominal)  Size implies order so nominal is misleading.
+violation(size_nominal) :-
+    attribute((scale,channel),E,size),
+    attribute((scale,type),E,categorical).
+
+% @hard(size_negative) Do not use size when data is negative as size implies that data is positive.
+violation(size_negative) :-
+    attribute(channel_field,_,size,N),
```

Comment:
This is hard to read. Why does `channel_field` have four components? It's unclear why we have `_` here. 

Note that creating unnecessary facts (especially ones with many components) can explode in space since the grounder has to instantiate many options.